internal/services/contracts_events_consumer_test.go

```diff
@@ -0,0 +1,256 @@
+)
+
+type mockTestEntity struct {
+	Contract    []byte
+	TokenID     types.Decimal
+	PrivilegeID int64
+	UserAddress []byte
+	ExpiresAt   time.Time
+}
+
+type mockTestArgs struct {
+	contract    common.Address
+	tokenID     types.Decimal
+	userAddress common.Address
+	expiresAt   int64
+	privilegeID int64
+}
+
+type cEventsTestHelper struct {
+	logger    zerolog.Logger
+	pdb       db.Store
+	container testcontainers.Container
+	ctx       context.Context
+	t         *testing.T
+	assert    *assert.Assertions
+}
+
+type eventsFactoryResp struct {
+	args    mockTestArgs
+	payload string
+}
+
+func TestProcessContractsEventsMessages(t *testing.T) {
+	s := initCEventsTestHelper(t)
+	defer s.destroy()
+
+	e := eventsPayloadFactory(1, 1, "", 0)
+	factoryResp := e[0]
+
+	msg := &message.Message{
+		Payload: []byte(factoryResp.payload),
+	}
+	c := NewContractsEventsConsumer(s.pdb, &s.logger)
+
+	err := c.processMessage(msg)
+	s.assert.NoError(err)
+
+	args := factoryResp.args
+
+	nft, err := models.FindNFTPrivilege(s.ctx, s.pdb.DBS().Reader, args.contract.Bytes(), args.tokenID, args.privilegeID, args.userAddress.Bytes())
+	s.assert.NoError(err)
+
+	s.assert.NotNil(nft)
+
+	actual := mockTestEntity{
+		Contract:    nft.ContractAddress,
+		TokenID:     nft.TokenID,
+		PrivilegeID: nft.Privilege,
+		UserAddress: nft.UserAddress,
+		ExpiresAt:   nft.Expiry,
+	}
+
+	expected := mockTestEntity{
+		Contract:    args.contract.Bytes(),
+		UserAddress: args.userAddress.Bytes(),
+		TokenID:     args.tokenID,
+		ExpiresAt:   time.Unix(args.expiresAt, 0).UTC(),
+		PrivilegeID: args.privilegeID,
+	}
+
+	s.assert.Equal(expected, actual, "Event was persisted properly")
+}
+
+func TestIgnoreWrongEventNames(t *testing.T) {
+	s := initCEventsTestHelper(t)
+	defer s.destroy()
+
+	e := eventsPayloadFactory(2, 2, "SomeEvent", 0)
+	factoryResp := e[0]
+
+	msg := &message.Message{
+		Payload: []byte(factoryResp.payload),
+	}
+	c := NewContractsEventsConsumer(s.pdb, &s.logger)
+
+	err := c.processMessage(msg)
+	s.assert.NoError(err)
+
+	s.assert.Nil(err)
+
+	args := factoryResp.args
+
+	nft, err := models.FindNFTPrivilege(s.ctx, s.pdb.DBS().Reader, args.contract.Bytes(), args.tokenID, args.privilegeID, args.userAddress.Bytes())
+	s.assert.EqualError(err, "sql: no rows in result set")
+
+	s.assert.Nil(nft)
+}
+
+func TestUpdatedTimestamp(t *testing.T) {
+	s := initCEventsTestHelper(t)
+	defer s.destroy()
+
+	e := eventsPayloadFactory(3, 3, "", 0)
+	factoryResp := e[0]
+
+	c := NewContractsEventsConsumer(s.pdb, &s.logger)
+
+	msg := &message.Message{
+		Payload: []byte(factoryResp.payload),
+	}
+	err := c.processMessage(msg)
+	s.assert.NoError(err)
+
+	args := factoryResp.args
+
+	oldNft, err := models.FindNFTPrivilege(s.ctx, s.pdb.DBS().Reader, args.contract.Bytes(), args.tokenID, args.privilegeID, args.userAddress.Bytes())
+	s.assert.NoError(err)
+
+	s.assert.NotNil(oldNft)
+
+	expiry := time.Now().Add(time.Hour + time.Duration(4)).UTC().Unix()
```

Comment:
Okay, I think I see the problem—none of these timestamps are mattering. The unit of `time.Duration` is  nanosecond, and `.Unix()` gets you to seconds.

The helper is making it very difficult to understand what's happening, but if you print out `string(msg.Payload)` in both instances you should see the same value for `expiry`.

Reply:
Yes, the scary thing is that if I remove `cold.Expiry` from the whitelist, this still passes!